File: benchmarks/bench_exceptions.js
#!/usr/bin/env node
// Node counterpart of bench_exceptions.strada.
class MiscErr extends Error {}
class WantedErr extends Error {}

let t0 = Date.now();
let ok = 0;
for (let i = 0; i < 2000000; i++) { try { ok++; } catch (e) { ok--; } }
let t1 = Date.now();
console.log("try-nothrow:", ok, (t1 - t0) / 1000);

let caught = 0;
for (let i = 0; i < 200000; i++) { try { throw new Error("boom " + i); } catch (e) { caught++; } }
let t2 = Date.now();
console.log("throw-catch:", caught, (t2 - t1) / 1000);

let typed = 0;
for (let i = 0; i < 200000; i++) {
    try { throw new WantedErr(); }
    catch (e) {
        if (e instanceof MiscErr) typed--;
        else if (e instanceof WantedErr) typed++;
        else typed--;
    }
}
let t3 = Date.now();
console.log("typed:", typed, (t3 - t2) / 1000);

function deep(n) { if (n <= 0) throw new Error("bottom"); return deep(n - 1) + 1; }
let deepCaught = 0;
for (let i = 0; i < 20000; i++) { try { deep(50); } catch (e) { deepCaught++; } }
let t4 = Date.now();
console.log("deep-unwind:", deepCaught, (t4 - t3) / 1000);

let fin = 0;
for (let i = 0; i < 1000000; i++) { try { fin++; } finally { fin++; } }
let t5 = Date.now();
console.log("finally:", fin, (t5 - t4) / 1000);
console.log("total:", (t5 - t0) / 1000);

File: benchmarks/bench_exceptions.pl
#!/usr/bin/env perl
# Perl counterpart of bench_exceptions.strada (eval/die; typed catch is
# emulated with ref() checks, the idiomatic Perl equivalent).
use strict; use warnings; use Time::HiRes qw(time);

my $t0 = time;
my $ok = 0;
for (1..2000000) { eval { $ok++ }; $ok-- if $@ }
my $t1 = time;
printf "try-nothrow: %d %.6f\n", $ok, $t1 - $t0;

my $caught = 0;
for my $i (1..200000) { eval { die "boom $i\n" }; $caught++ if $@ }
my $t2 = time;
printf "throw-catch: %d %.6f\n", $caught, $t2 - $t1;

my $typed = 0;
for (1..200000) {
    eval { die bless {}, 'WantedErr' };
    if (ref $@ eq 'MiscErr') { $typed-- }
    elsif (ref $@ eq 'WantedErr') { $typed++ }
    else { $typed-- }
}
my $t3 = time;
printf "typed: %d %.6f\n", $typed, $t3 - $t2;

sub deep { my $n = shift; die "bottom\n" if $n <= 0; return deep($n - 1) + 1 }
my $deep_caught = 0;
for (1..20000) { eval { deep(50) }; $deep_caught++ if $@ }
my $t4 = time;
printf "deep-unwind: %d %.6f\n", $deep_caught, $t4 - $t3;

my $fin = 0;
for (1..1000000) { eval { $fin++ }; $fin++ }   # eval + unconditional cleanup
my $t5 = time;
printf "finally: %d %.6f\n", $fin, $t5 - $t4;
printf "total: %.6f\n", $t5 - $t0;
